core/src/main/java/org/elasticsearch/index/fielddata/IndexFieldDataService.java

```diff
@@ -115,18 +116,20 @@ public void clearField(final String fieldName) {
         IndexFieldData.Builder builder = fieldType.fielddataBuilder(fullyQualifiedIndexName);
 
         IndexFieldDataCache cache;
-        cache = fieldDataCaches.get(fieldName);
-        if (cache == null) {
-            String cacheType = indexSettings.getValue(INDEX_FIELDDATA_CACHE_KEY);
-            if (FIELDDATA_CACHE_VALUE_NODE.equals(cacheType)) {
-                cache = indicesFieldDataCache.buildIndexFieldDataCache(listener, index(), fieldName);
-            } else if ("none".equals(cacheType)){
-                cache = new IndexFieldDataCache.None();
-            } else {
-                throw new IllegalArgumentException("cache type not supported [" + cacheType + "] for field [" + fieldName + "]");
+        cache = fieldDataCaches.computeIfAbsent(fieldName,
+            s -> {
```

Comment:
It might be cleaner and create less new Function objects if you extract this compute block as a new method, say "newIndexFieldDataCache(fieldName)", then just do `fieldDataCaches.computeIfAbsent(fieldName, this::newIndexFieldDataCache)` here.

Reply:
added `synchronized` back